jedi/inference/gradual/typing.py

```diff
@@ -143,6 +143,42 @@ def _create_instance_with_generics(self, generics_manager):
             generics_manager
         )
 
+    def infer_type_vars(self, value_set, is_class_value=False):
+        # Circular
+        from jedi.inference.gradual.annotation import merge_pairwise_generics, merge_type_var_dicts
+
+        type_var_dict = {}
+        annotation_generics = self.get_generics()
+
+        if not annotation_generics:
+            return type_var_dict
+
+        annotation_name = self.py__name__()
+        if annotation_name == 'Optional':
+            if is_class_value:
+                # This only applies if we are comparing something like
+                # List[Optional[int]] with Iterable[Optional[int]]. First, Jedi
+                # tries to match List/Iterable. After that we will land here,
+                # because is_class_value will be True at that point. Obviously
+                # we also compare below that both sides are `Optional`.
+                for element in value_set:
+                    element_name = element.py__name__()
+                    if element_name == 'Optional':
```

Comment:
It's since occurred to me that this isn't strictly correct -- it's completely valid to pass a non-optional value into a function which expects an optional one. I'll need to add some tests for this to see if this works for that case, though I suspect it won't.

There is a slight gotcha here, in that `mypy` won't actually allow this in all cases due to covariance rules. For example, you can't pass `List[int]` to `List[Optional[int]]` even though you can pass it to `Sequence[Optional[int]]`.

I _suspect_ jedi might just be able to ignore that though. It'll mean jedi will offer completions where `mypy` might complain about the types, but I'm not sure if that's a big deal. (IMO that's preferable to not offering completions where they could be valid).

Reply:
> mypy won't actually allow this in all cases due to covariance rules

Hmm why is that? Isn't the former a covariant of the latter and should therefore be OK? It feels quite stupid to pass that. (I'm assuming we are talking about a case where you pass the former to `def foo(x: List[Optional[int]]): ...`).

> I suspect jedi might just be able to ignore that though. It'll mean jedi will offer completions where mypy might complain about the types, but I'm not sure if that's a big deal. (IMO that's preferable to not offering completions where they could be valid).

100% agree. Mypy and Jedi have different goals. I also think that the case you mention would never result in an issue, because it's valid typing.

Reply:
Covariance rules are a little complicated to grok. Essentially it comes down to this sort of behaviour:

``` python
def cannot_accept_non_optional(items: List[Optional[int]]) -> None:
    items.append(None)

def can_accept_non_optional(items: Sequence[Optional[int]]) -> None:
    # Here items is immutable (from a type system perspective), so while
    # the code inside the function needs to cope with `None`s, they can't
    # leak out
    print(len(items))

items = [42]  # type: List[int]

cannot_accept_non_optional(items) # bad
can_accept_non_optional(items) # ok

for item in items:
    print(item * 2)
```

Reply:
Oh, thanks a lot for these examples! I understand. I just never even thought of passing a mutable list :). I don't really write code like this, so I guess the rule makes sense if it's not immutable. It's even invariant as far as I can see.

Reply:
> It's since occurred to me that this isn't strictly correct -- it's completely valid to pass a non-optional value into a function which expects an optional one. I'll need to add some tests for this to see if this works for that case, though I suspect it won't.

7c6c62a5 adds said tests. After updating this branch for the recent removal of `is_class_value` (i.e; removing that branch) it turns out that those tests already pass :)